GSdx HW: Add Skipdraw Offset option for Skipdraw hack (continued)

turtleli opened the pull request:
Continuation of #2290.

I added Linux GUI code and simplified some of the Windows GUI code and misc other stuff.

ivan89el commented:
I have no problems, it works as well https://github.com/PCSX2/pcsx2/pull/2290
(I have windows 7 x64)

turtleli commented:
Updated the Linux GUI so it says Skipdraw Range on Linux as well.

ivan89el commented:
I found a problem: I do not save the first value and always reset to 1

turtleli commented:
Should be fixed now.

ivan89el commented:
Now everything is good.

lightningterror commented on plugins/GSdx/GSSetting.cpp:
```diff
@@ -57,9 +57,13 @@ const char* dialog_message(int ID, bool* updateText) {
+		case IDC_SKIPDRAWOFFSETEDIT:
+			return "Completely skips drawing surfaces from the surface in the left box up to the surface specified in the box on the right.\n\n"
+				"Use it, for example, to try and get rid of bad post processing effects.\n"
+				"Step 1: Increase the value in the left box and keep the value in the right box set to the same value as the left box to find and remove a bad effect. Try values between 1 and 100.\n"
```
The gui allows values up to 10000.

turtleli replied:
I read it as a recommendation. Alternatively it could be removed.

BTW, you approved #2290 but the wording here hasn't changed. ;)

lightningterror replied:
>I read it as a recommendation. Alternatively it could be removed.

Ye I kinda guessed that, I say we remove it.

>BTW, you approved #2290 but the wording here hasn't changed. ;)

Thought there was no need for that since you're the one PRing it :)

turtleli replied:
> Thought there was no need for that since you're the one PRing it :)

My point was that you've kinda approved something that you want changes to be made to, in which case it's ambiguous as to whether you've approved it or not.

lightningterror replied:
Ah, well in this case it's just a miniature nitpick but I'm fine with it being merged either way.

turtleli commented:
Updated to address review comment.

lightningterror commented on plugins/GSdx/GSdx.rc:
```diff
@@ -154,24 +154,26 @@ BEGIN
+    RTEXT           "Geometry Shader:",IDC_GEOMETRY_SHADER_TEXT,14,241,62,8
+    COMBOBOX        IDC_GEOMETRY_SHADER_OVERRIDE,80,238,97,63,CBS_DROPDOWNLIST | WS_VSCROLL | WS_TABSTOP
+    RTEXT           "Image Load Store:",IDC_IMAGE_LOAD_STORE_TEXT,14,256,62,8
+    COMBOBOX        IDC_IMAGE_LOAD_STORE,80,253,97,63,CBS_DROPDOWNLIST | WS_VSCROLL | WS_TABSTOP
```
Since you're changing gl hacks maybe you can adjust IDC_GEOMETRY_SHADER_OVERRIDE, IDC_IMAGE_LOAD_STORE width to match HPO, Sprite ..etc

turtleli commented:
Updated to address review comment.

lightningterror commented on plugins/GSdx/GSSetting.cpp:
```diff
@@ -57,9 +57,13 @@ const char* dialog_message(int ID, bool* updateText) {
+				"Use it, for example, to try and get rid of bad post processing effects.\n"
+				"Step 1: Increase the value in the left box and keep the value in the right box set to the same value as the left box to find and remove a bad effect.\n"
+				"Step 2: If a bad effect found with Step 1 is not completely removed yet, then without changing the value in the left box, try increasing the value in the box to right until the effect is completely gone.\n";
+				"Note: Increase the value in the right box and keep the value in the left box set to \"1\" to reproduce the old skipdraw behaviour.";
```
I don't see the last line in the tooltip, maybe because of \"1\" ?

turtleli replied:
It's the `;`

lightningterror replied:
Ah right, didn't notice that.

turtleli commented:
Updated to improve value correction and fix the tooltip.

lightningterror commented:
Sweet, it works much better now. Is there anything else that needs tweaking ?
If not then this could be ready for a merge.

lightningterror commented:
Once this gets merged I'm also thinking of adjusting the TC Offsets in to one line. Kinda like this new skipdraw feature.

TC Offset X,Y - X on the left and Y on the right.

turtleli commented:
Updated to fix the comment (I mixed up skipdraw and skipdraw offset in one place). No functional changes.

turtleli commented:
Updated again. I changed the Linux behaviour back to what I had at the start of the PR - it doesn't have the same issue Windows does.

> Uh, I had always understood this PR to be merged *after* flatout one, not replace it

It's easier to just replace it in this case.

mirh commented:
Obviously, sure. 
But I can remember somebody kind-of-complaining in the past for authorship vs commiting, or something.

turtleli commented:
I set the commit author to FlatOut.
```
git commit --author="namename<emailemail>"
```

lightningterror commented:
> Updated again. I changed the Linux behaviour back to what I had at the start of the PR - it doesn't have the same issue Windows does.

Seems a bit odd that it doesn't.

lightningterror commented:
@turtleli is this ready for a merge ?

turtleli commented:
Probably? I'll merge it after some (hopefully) final checks.

> Seems a bit odd that it doesn't.

It's not exactly surprising if the OS and toolkit are both different.

turtleli commented on plugins/GSdx/GSHwHack.cpp:
```diff
@@ -2364,7 +2364,8 @@ bool GSState::IsBadFrame()
 			if (GSLocalMemory::m_psm[fi.TPSM].depth || GSUtil::HasSharedBits(fi.FBP, fi.FPSM, fi.TBP0, fi.TPSM))
 			{
-				m_skip = m_userhacks_skipdraw;
+				m_skip_offset = std::max(m_userhacks_skipdraw_offset, 1);
```
The `std::max` is unnecessary it seems.

lightningterror commented:
> It's not exactly surprising if the OS and toolkit are both different.

But isn't it better to keep that code functionality even if the issue is not present ?

turtleli commented:
Keep what functionality?

lightningterror commented:
> Keep what functionality?

> Updated again. I changed the Linux behaviour back to what I had at the start of the PR - it doesn't have the same issue Windows does.

Keeping the updated version and not how it was at the start of the PR.

lightningterror commented:
@turtleli I made some adjustments to the TC offsets hack and put them in one line like skipdraw offset. Do you wanna include these changes as well if they are alright ?

![image](https://user-images.githubusercontent.com/18107717/44343667-e3441000-a48e-11e8-92e4-f7f3c06455bf.png)

https://github.com/lightningterror/pcsx2/commit/dc9ad1dc526a5eec35d00840f61a8c3247d61f35

turtleli commented:
I'd prefer not to include it. It's unrelated to the PR and I'd have to delay merging this PR (which will probably be done tonight) so people have time to give feedback on those changes.

lightningterror commented:
Sure thing, I'll look in to adding them in the future once this gets merged then.

turtleli commented:
...Updated again because I accidentally undid the Linux changes on my previous push (was on Windows and didn't realise my local branch was out of sync). That's fixed now.

Will probably still merge it once the buildbots have finished running.
